fl4health/privacy_mechanisms/dpPoissonBinomial.py

```diff
@@ -0,0 +1,162 @@
+
+    Reference
+        The Poisson Binomial Mechanism for Unbiased Federated Learning with Secure Aggregation
+        https://proceedings.mlr.press/v162/chen22s/chen22s.pdf
+    """
+
+    if N < 1 or type(N) != int:
+        raise Exception("The parameter N of Binom(N, p) must be a natural number")
+
+    if radius < 0:
+        raise Exception("The radius for the interval must be positive")
+
+    if not (0 <= theta <= 0.25):
+        raise Exception("theta must be within the real interval [0, 1/4]")
+
+    # rescaling
+    p = theta * x / radius + 1 / 2
+
+    # privatization
+    B = Binomial(N, p)
+
+    return B.sample().int().item()
+
+
+def Kashin_alternative(dim: int) -> List[List[int]]:
+    """
+    Server generated random rotation matrix that replaces.
+
+    Args:
+        dim (int): Dimension.
+
+    Returns:
+        List[List[int]]: Replacement of the tight frame in Kashin representation.
+    """
+
+    sample_matrix = ortho_group(dim).rvs()
+
+    return sample_matrix.tolist()
+
+
+def PoissonBinomialMechanism_Client(
+    query_vector: List[float], radius: float, tight_frame: List[float], theta: float, K: float, modulus: int
+) -> List[int]:
+    """
+    Privatization of client vector with Poisson bionomial mechanism.
+
+    Args:
+        query_vector (List[float]): Client integer vector.
+        radius (float): The radius.
+        tight_frame (List[float]): Matrix to compute Kashin representation.
+        theta (float): A parameter within the real interval [0, 1/4].
+        K (float): Constant parameter in Kashin representation, or clip.
+        modulus (int): The modulus for modular arithmetic used in SecAgg.
+
+    Raises:
+        Exception: Bad query.
+
+    Returns:
+        List[int]: The discretized and privatized query_vector as a list of integers.
+
+    Reference
+        The Poisson Binomial Mechanism for Unbiased Federated Learning with Secure Aggregation
+        https://proceedings.mlr.press/v162/chen22s/chen22s.pdf
+    """
+
+    dim = len(query_vector)
+
+    if dim == 0:
+        raise Exception("query_vector cannot be empty")
+
+    kashin_matrix = torch.tensor(tight_frame)
+    client_vector = torch.tensor(query_vector)
+
+    kashin_vector = torch.matmul(kashin_matrix, client_vector)
+
+    # TODO adjust this for l-infinity clipping
+    scalar_radius = radius * K / math.sqrt(dim)
+
+    privatization = torch.zeros(dim)
+    for j in range(dim):
+        privatization[j] += ScalerPoissonBinomial(kashin_vector[j].item(), modulus - 1, scalar_radius, theta)
+
+    return list(privatization)
+
+
+def PoissonBinomialMechanism_Server(
+    inverse_matrix: List[List[int]], radius: float, modulus: int, theta: float, *client_vectors: List[int]
+) -> List[int]:
+    """
+    Server side function that recovers the sum of vectors from clients.
+
+    Args:
+        inverse_matrix (List[List[int]]): The inverse of the random rotation generated by Kashin_alternative().
+        radius (float): Radius.
```

Comment:
A bit more detail would be useful here. It appears that this should be the scaled version of the radius if I'm reading the algorithm correctly. That is, radius should be c' = \frac{cK}{\sqrt{d}}?